### PR TITLE
Sue Creek: draft initial testing port

### PR DESCRIPTION
**Pull request by lyakh**
This is more of an initial indication and a review request, than a pull request. Status: with a bit of manual image manipulation this boots on Sue Creek.

### CODE REVIEW

File: src/drivers/intel/dw-uart.c [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	uint32_t timeout;
+};
+
+static struct dw_uart_device uart_dev = {
+	.port = SUE_UART_REG_BASEADDR,
+	.timeout = SUE_UART_TIMEOUT,
+};
+
+
+void dw_uart_init(uint32_t baud)
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: It probably makes sense to make this API more generic. Linux has the concept of low level debug before the console is ready, we could have the same i.e. ll_dbg_uart_init(baud, bits, stop, parity)

File: src/drivers/intel/dw-uart.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+{
+	struct dw_uart_device *dev;
+	uint32_t divisor, tmp;
+
+	dev = (struct dw_uart_device *)&uart_dev;
+
+	if (baud != 0) {
+		divisor = (SUE_SYS_CLK_FREQ / baud) >> 4;
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: SUE_SYS_CLK_FREQ would be defined in platform.h as generic PLATFORM_LL_UART_CLK_FREQ, same goes for other platform details.

File: src/drivers/intel/dw-uart.c [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		uart_write(dev, SUE_UART_REG_BRDL,
+					(uint8_t)(divisor & 0xFF));
+		uart_write(dev, SUE_UART_REG_BRDH,
+					(uint8_t)((divisor >> 8) & 0xFF));
+
+		/* restore to access baudrate divisor registers */
+		uart_write(dev, SUE_UART_REG_LCR, tmp);
+	}
+
+	/* 8Bit-data, 1-stop bit, no parity, clear DLAB */
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: This would be passed in by bootldr.

File: src/drivers/intel/dw-uart.c [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	/* FIFO enalbe, mode0, Tx/Rx FIFO reset */
+	uart_write(dev, SUE_UART_REG_FCR,
+		FCR_FIFOE(1) | FCR_MODE(0) | FCR_RCVR_RST | FCR_XMIT_RST);
+
+	/* reset port */
+	uart_write(dev, SUE_UART_REG_RBR, 0);
+}
+
+
+void dw_uart_write_word(uint32_t word)
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: I think we need ll_dbg_uart_write_word()

File: src/drivers/intel/dw-uart.c [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+{
+	struct dw_uart_device *dev;
+	uint8_t bytes[9];
+	uint32_t outchar;
+	int i, j;
+	uint32_t timeout;
+
+	dev = (struct dw_uart_device *)&uart_dev;
+
+	for (i = 28, j = 0; i >= 0 && j < 8 ; i -= 4, j++)
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: This needs comment explaining what it does and macros for the magic numbers.

File: src/include/sof/dw-uart.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Zhigang Wu <zhigang.wu@intel.com>
+ */
+#ifndef __INCLUDE_DW_UART_H__
+#define __INCLUDE_DW_UART_H__
+
+#define SUE_SYS_CLK_FREQ		38400000
+#define SUE_UART_REG_BASEADDR	0x00080800
+#define SUE_UART_TIMEOUT		25000
```
Comment (lgirdwood) [inline src/include/sof/dw-uart.h]: base addr should be in memory.h, freq should be in platform.h alongside timeout.

File: src/drivers/intel/dw-uart.c [hunk 7]
```diff
@@ -160,3 +175,113 @@ void dw_uart_write_word(uint32_t word)
...
+/* Actually the FIFO size can be read out */
+
+#define DW_UART_RING_SIZE 32
+static uint8_t dw_uart_ring[DW_UART_RING_SIZE];
+/* write to head, read from tail */
+static bool dw_uart_ring_empty = true;
+static unsigned int dw_uart_ring_head;
+static unsigned int dw_uart_ring_tail;
+
+#define min(a, b) ({		\
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: I think there is already a min() in sof.h (if not, this is where to add it).

File: src/drivers/intel/dw-uart.c [hunk 8]
```diff
@@ -160,3 +175,113 @@ void dw_uart_write_word(uint32_t word)
...
+		size -= count;
+	}
+
+	dw_uart_ring_empty = false;
+
+	return size;
+}
+
+/* Block until all the data is at least in the ring buffer */
+int dw_uart_write(const uint8_t *data, int size)
```
Comment (lgirdwood) [inline src/drivers/intel/dw-uart.c]: so for baseFW users we can pass in a dev * here. See ssp.c and dw-dma.c (they also have probe() and remove()).

**Comment by sofci** [comment]
Can one of the admins verify this patch?

File: src/drivers/intel/dw-uart.c [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	uint32_t baud;	/* Baud rate */
+	uint32_t timeout;
+};
+
+static struct dw_uart_device uart_dev = {
+	.port = SUE_UART_REG_BASEADDR,
+	.timeout = SUE_UART_TIMEOUT,
+};
+
+
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/drivers/intel/dw-uart.c]: Could you please clarify this a bit. The only code that is only called from the boot loader is the UART initialisation. The dw_uart_write_word() function, used to implement trace_point(), is used from both. The UART code, used for printf()-like trace functionality is only used from the FW proper, Are you proposing to put

boot loader (dw-uart-ll):
UART init
trace_point()

FW proper:
trace_point()
trace_event()

? Further, for trace_event() we need an sprintf() implementation in SOF. AFAICS we only support integer parameters (%d, %p, %u, %x) and only up to 4 of them. So, I could code something myself or we could copy some BSD-licensed implementation?

File: src/drivers/intel/dw-uart.c (hunk 1 above)
Comment (lyakh) [reply src/drivers/intel/dw-uart.c]: I found
static void early_serial_init(int port, int baud)
on x86 (on ARM early printk seems to rely on the bootloader UART configuration), is that what you meant? Do we really need stop bits and parity?..

**Comment by lgirdwood** [comment]
Can you rebase and retest on top of #740

File: src/drivers/intel/dw-uart.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/drivers/intel/dw-uart.c]: I mean this should be like linux where we have a very simple low level UART driver (with no resource/PM management) that only exports one or two APIs and a more comprehensive UART driver (that does resource mgmt, PM, IRQ etc). The LL driver would poll/block and only be used by bootloader code, whilst the full driver used by general base FW.

So yes, your example is correct, but we dont need sprintf as this is already done, we just need to write binary data out of UART (and pipe it to sof-logger).

e.g.
```sof-logger -u /dev/ttyUSB0```

File: src/drivers/intel/dw-uart.c (hunk 1 above)
Comment (lgirdwood) [reply src/drivers/intel/dw-uart.c]: yes, need stop bits and parity, easy to implement. They can be hard coded by platforms to 8N1 but they should be there for completeness

Comment (lyakh) [reply src/drivers/intel/dw-uart.c]: Can we use
uart_ll_init(baud,...) - called from the boot-loader
uart_init(baud,...) - called from SOF
uart_write_word() - used by both
uart_write() - SOF
uart_write_nowait() - SOF?

**Comment by lgirdwood** [comment]
Will close this since it's being dealt with in multiple PRs.

**Pull request closed without merging**